components/auth/org.wso2.carbon.auth.oauth/src/main/java/org/wso2/carbon/auth/oauth/impl/JWTTokenGenerator.java

```diff
@@ -0,0 +1,230 @@
+                context.getParams().put(OAuthConstants.TOKEN, jwtToken);
+                super.generateAccessToken(context);
+            } catch (AuthException e) {
+                context.setSuccessful(false);
+                context.setErrorObject(OAuth2Error.SERVER_ERROR);
+            }
+        }
+    }
+
+    @Override
+    public boolean renewAccessTokenPerRequest() {
+
+        return true;
+    }
+
+    private String signJwt(JWTClaimsSet jwtClaimsSet) throws AuthException {
+
+        if (JWSAlgorithm.RS256.equals(signatureAlgorithm) || JWSAlgorithm.RS384.equals(signatureAlgorithm) ||
+                JWSAlgorithm.RS512.equals(signatureAlgorithm)) {
+            try {
+                JWSSigner signer = new RSASSASigner(ServiceReferenceHolder.getInstance().getPrivateKey());
+                JWSAlgorithm jwsAlgorithm;
+                if (signatureAlgorithm instanceof JWSAlgorithm) {
+                    jwsAlgorithm = (JWSAlgorithm) signatureAlgorithm;
+                } else {
+                    throw new AuthException("Signature Algorithm couldn't convert to JWSAlgorithm");
+                }
+                JWSHeader.Builder headerBuilder = new JWSHeader.Builder(jwsAlgorithm);
+                String certThumbPrint = getThumbPrint(ServiceReferenceHolder.getInstance().getPublicKey());
+                headerBuilder.keyID(certThumbPrint);
+                headerBuilder.x509CertThumbprint(new Base64URL(certThumbPrint));
+                SignedJWT signedJWT = new SignedJWT(headerBuilder.build(), jwtClaimsSet);
+                signedJWT.sign(signer);
+                return signedJWT.serialize();
+            } catch (NoSuchAlgorithmException | CertificateEncodingException | JOSEException e) {
+                throw new AuthException("Invalid signature algorithm provided. " + signatureAlgorithm);
+            }
+        } else {
+            throw new AuthException("Invalid signature algorithm provided. " + signatureAlgorithm);
+        }
+    }
+
+    private JWTClaimsSet getJWTClaimSet(AccessTokenContext context) {
+
+        Scope scope = (Scope) context.getParams().get(OAuthConstants.SCOPES);
+        long defaultValidityPeriod = (long) context.getParams().get(OAuthConstants.VALIDITY_PERIOD);
+        long curTimeInMillis = Calendar.getInstance().getTimeInMillis();
+        String sub = (String) context.getParams().get(OAuthConstants.AUTH_USER);
+        String consumerKey = (String) context.getParams().get(OAuthConstants.CLIENT_ID);
+        String issuer = ServiceReferenceHolder.getInstance().getAuthConfigurations().getTokenIssuer();
+        JWTClaimsSet.Builder jwtClaimsSetBuilder = new JWTClaimsSet.Builder();
+        jwtClaimsSetBuilder.issuer(issuer);
+        jwtClaimsSetBuilder.subject(sub);
+        jwtClaimsSetBuilder.claim(AUTHORIZATION_PARTY, consumerKey);
+        jwtClaimsSetBuilder.issueTime(new Date(curTimeInMillis));
```

Comment:
This value should be in seconds, not in milli seconds

Reply:
The above line converting value to seconds